pyvelox/substrait.cpp

```diff
@@ -0,0 +1,237 @@
+  std::ifstream msgJson(msgPath);
+  VELOX_CHECK(
+      !msgJson.fail(), "Failed to open file: {}. {}", msgPath, strerror(errno));
+  std::stringstream buffer;
+  buffer << msgJson.rdbuf();
+  std::string msgData = buffer.str();
+  auto status = google::protobuf::util::JsonStringToMessage(msgData, &msg);
+  VELOX_CHECK(
+      status.ok(),
+      "Failed to parse Substrait JSON: {} {}",
+      status.code(),
+      status.message());
+}
+
+static inline void initializeSubstrait() {
+  PySubstraitContext::getInstance().initialize();
+}
+
+static inline void finalizeSubstrait() {
+  PySubstraitContext::getInstance().finalize();
+}
+
+std::vector<std::shared_ptr<facebook::velox::connector::ConnectorSplit>>
+makeSplits(
+    const facebook::velox::substrait::SubstraitVeloxPlanConverter& converter,
+    const std::shared_ptr<const core::PlanNode> planNode,
+    const std::string& dirPath) {
+  const auto& splitInfos = converter.splitInfos();
+  auto leafPlanNodeIds = planNode->leafPlanNodeIds();
+  // Only one leaf node is expected here.
+  EXPECT_EQ(1, leafPlanNodeIds.size());
+  const auto& splitInfo = splitInfos.at(*leafPlanNodeIds.begin());
+
+  const auto& paths = splitInfo->paths;
+  const auto& starts = splitInfo->starts;
+  const auto& lengths = splitInfo->lengths;
+  const auto fileFormat = splitInfo->format;
+
+  std::vector<std::shared_ptr<facebook::velox::connector::ConnectorSplit>>
+      splits;
+  splits.reserve(paths.size());
+
+  for (int i = 0; i < paths.size(); i++) {
+    auto path = fmt::format("{}{}", dirPath, paths[i]);
+    auto start = starts[i];
+    auto length = lengths[i];
+    auto split = facebook::velox::exec::test::HiveConnectorSplitBuilder(path)
+                     .fileFormat(fileFormat)
+                     .start(start)
+                     .length(length)
+                     .build();
+    splits.emplace_back(split);
+  }
+  return splits;
+}
+
+static inline RowVectorPtr runSubstraitQuery(
+    const std::string& plan,
+    bool enableSplits,
+    const std::string& dirPath) {
+  memory::MemoryPool* pool = PyVeloxContext::getSingletonInstance().pool();
+
+  facebook::velox::substrait::SubstraitVeloxPlanConverter planConverter(pool);
+
+  ::substrait::Plan substraitPlan;
+  // convert plan to protobuf
+  // read from file
+  if (isJSON(plan)) {
+    readFromJSON(plan, substraitPlan);
+  } else if (isFilePath(plan)) {
+    // check whether input file is in JSON format
+    if (isJsonFile(plan)) {
+      readFromFile(plan, substraitPlan);
+    } else {
+      /// other formats are not yet supported.
+      /// NOTE: To support Protobuf format, we should generate Substrait proto
+      /// in Python format too. Then we should be able to add proto format too.
+      throw py::value_error("plan should be path to a plan in JSON format.");
+    }
+  } else {
+    throw py::value_error("Invalid Substrait plan.");
+  }
+
+  auto planNode = planConverter.toVeloxPlan(substraitPlan);
```

Comment:
Will this always succeed ?

Reply:
Good point, it could definitely fail if we choose unsupported features (Velox end). Should we rather use `VELOX_CHECK`?